### ISSUE TITLE
Follow-ups from the workflow RNG system (#400)

### ISSUE THREAD

**Issue by jhuggins** [TARPS-group/prob-pipe#429]
## Context

PR #400 landed the workflow-owned structured RNG system from #389: `workflow_run(seed=...)`,
structural key derivation from (workflow root, occurrence path, stochastic source id, logical
unit id), a per-call broker, co-sampling, managed execution with retry claims across
local/thread/Prefect routes, provenance `controls`/`diagnostics`, and `replay_run(provenance)`.

These are the items surfaced during review of that branch and deliberately left out to keep the
PR reviewable. Nothing here is a regression in shipped behavior, and the RNG contract itself
verified sound — the escape path, replay exactness, and cross-route key identity are all covered
by tests. Item 1 is the only one with a correctness consequence; the rest are hardening, cleanup,
docs, and test coverage.

Line numbers are against `main` at `12c642fd`. Findings come from a Claude-assisted review of the
merged branch.

## 1. Partial managed registration masks the original exception

Both managed entry points register work items *before* the guarded region:

| Function | Registration | `try` |
| --- | --- | --- |
| `execute_many` | `probpipe/core/_workflow_execution.py:124` | `:126` |
| `execute_many_threaded` | `:179` | `:214` |

If `register_managed_work_items` raises partway — for example a duplicate work-item token across
items — the units already registered stay `ISSUED`, and the `finally` that would join or cancel
them never runs. The broker scope's failure path
(`probpipe/core/_workflow_broker.py:1133`, `except BaseException: broker._finalize(successful=False)`)
then reaches `_begin_finalization`, whose join check raises

```
a stochastic broker scope cannot exit before all managed work items join
```

at `probpipe/core/_workflow_broker.py:665`. Two consequences: the real cause is only reachable
through `__context__`, and because that raise happens *before* `self._lifecycle =
_BrokerLifecycle.FINALIZING` at `:668`, `_seal_finalization` never runs and the broker is left
`OPEN`.

Either move registration inside the guarded region, or have `_finalize(successful=False)` cancel
outstanding reservations rather than asserting they joined.

## 2. Retry schedule charges the batch the most-retried item's delay

`probpipe/core/_workflow_execution.py:418`:

```python
retry_ordinal = max(retries_by_index[outcome.index] for outcome in retryable)
```

Retry counts inside `retryable` can diverge, because `_coordinate_prefect_randomness` can turn a
previously successful outcome into an errored one. A fresh failure entering on round *k* is then
delayed by `delays[k-1]` instead of `delays[0]`, skipping the ramp. This is invisible with the
shipped scalar default and only shows with a list-valued
`PREFECT_TASKS_DEFAULT_RETRY_DELAY_SECONDS`. `min`, or a per-item delay, matches Prefect's
per-task-run semantics more closely.

Related: `_retry_delay_for_ordinal` (`:492`) indexes `delays[min(ordinal - 1, len(delays) - 1)]`,
so an ordinal of `0` would silently return the last delay. Callers currently increment first, but
the helper is standalone now and deserves an explicit guard.

## 3. `dispatch="jax"` gained an undocumented requirement

The route probe no longer samples — it builds `jax.ShapeDtypeStruct` values from each sampled
root's event template and traces `jax.vmap` over those, which is a real improvement over drawing
real samples just to build a jaxpr. It does mean every sampled root must now declare a concrete
`NumericEventTemplate` **and** `dtypes`, or the probe raises (`probpipe/core/node.py:1061`,
`:1068`):

```
<Type> does not declare a concrete numeric event template for side-effect-free JAX probing
```

Under `dispatch="auto"` that is a graceful fallback to row-wise execution. Under an explicit
`dispatch="jax"` it is a hard `TypeError`. `dtypes` is present on the main distribution families
(the TFP base class covers most concrete distributions, and `SequentialJointDistribution` gained
one in #400), so common cases are unaffected — but a custom distribution that worked under
`dispatch="jax"` before #400 can now fail, and nothing in the CHANGELOG or `docs/api/` says so.

Worth a CHANGELOG note and a line in the dispatch documentation stating what a root must expose
to be eligible for JAX dispatch.

## 4. Smaller code items

- **Lock per hop on the key-derivation path.** `_find_remote_coordination_observation`
  (`probpipe/core/_workflow_context.py:465`) takes `cursor.state.lock` on every frame in the
  chain, and it runs on every `key_for`, replacing what used to be a lock-free `ContextVar.get()`.
  Sibling workers in a thread route contend on the same common-ancestor lock that
  `_resolve_root_words` and `_materialize_path` also need. The field is a single attribute and the
  walk is not atomic anyway, so a plain read (or a memoized "chain has no observation" flag) would
  remove the contention.
- **`allow_nan=False` missing from one of three `_canonical_json` copies.**
  `probpipe/core/_workflow_recipe.py:282` still omits it; the copies in `_workflow_replay.py` and
  `_workflow_callable.py` set it. The descriptor route is now closed by
  `_validate_descriptor_value`, so the residual risk is narrow, but the asymmetry means a
  non-finite value reaching the recipe encoder would emit invalid JSON rather than failing.
- **PPC test-function uniqueness keys on `__name__`.**
  `probpipe/diagnostics/_ppc_spc.py:266` uses `getattr(fn, "__name__", None)` falling back to
  `repr(fn)`. Two bound methods of *different* instances share a `__name__` and are rejected
  despite being genuinely distinct, while callables without `__name__` pass via an address-bearing
  `repr` that makes result keys non-reproducible. `__qualname__` plus rejecting the address
  fallback would be tighter.
- **Two copies of one count validator.** `_validate_positive_int`
  (`probpipe/validation/_workflow_rng.py:16`) and `_validate_conversion_sample_count`
  (`probpipe/converters/_registry.py:37`) have identical bodies differing only in the message
  noun; the former already generalizes over the name.
- **Bare `assert` for an internal invariant.** `probpipe/converters/_scipy.py:295` has
  `assert sample_shape is not None`, which is stripped under `python -O`. The equivalent in
  `probpipe/validation/_comparison.py` was converted to a `RuntimeError` in #400, so this is now
  the odd one out.
- **Coordination teardown no longer checks identity.**
  `probpipe/core/_workflow_broker.py:1274` clears
  `frame.state.remote_coordination_observation = None` unconditionally. Harmless today because
  each transported payload allocates a fresh frame and observation, but incorrect nesting would
  silently clear another observation rather than fail.
- **Nested-path dtype fallback takes the parent's dtype.** `probpipe/core/node.py:1074` falls back
  to `dtypes[path.split("/", 1)[0]]` when a leaf path is absent. For a nested record whose leaf
  dtype differs from its top-level field's, the probe traces at the wrong dtype. It is only a
  probe, so the worst case is a spurious trace failure (which `auto` absorbs) or a probe that
  passes where real execution differs.

## 5. Test coverage and hygiene

- **`probpipe/core/_workflow_broker.py:852` is untested.** The guard
  `"a rootless remote coordination probe cannot report stochastic claims"` covers the late or
  illegal coordination-report path and has no test referencing it. The other new coordination
  invariants were either covered or removed as unreachable in #400; this one is reachable.
- **The end-to-end golden vector is fragile.** `tests/core/test_workflow_rng_lifting.py:81` pins
  `sample_calls == [((3974922193, 721833970), (5,))]` plus the resulting values as bare literals,
  with no regeneration note and no gate on the JAX version or `jax_default_prng_impl` — a
  non-threefry impl changes both the key arity and the bits. `tests/core/test_workflow_recipe.py`
  has the better in-repo idiom: goldens keyed by ABI with an actionable failure message.
- **The automatic-key source scan asserts exact source substrings.**
  `tests/core/test_automatic_key_source_scan.py:158-160` asserts strings such as
  `"self._key, key = jax.random.split(self._key)"` are present in `probpipe/modeling/_glm.py`, so
  an innocuous refactor there fails an RNG test. The scan also does not recognize
  `jax.random.wrap_key_data` or `getattr(jax.random, "key")`.
- **The `prefect` marker is declared but unused.** `pyproject.toml:79` declares it; no test uses
  `@pytest.mark.prefect`. `tests/core/test_prefect_orchestration.py` gates on a module-level
  `importorskip`, so `-m "not prefect"` deselects nothing. Either apply the marker to that module
  or drop the declaration.
- **A few weak assertions.** `assert result is not None` in `tests/core/test_workflow_parallel.py`
  cannot fail for a function returning a uint32 scalar; the test named
  `..._is_transactional_and_preserves_cause` makes no assertion about a cause (`pytest.raises`
  does not inspect `__cause__`), and its `SEALED` check runs after the `with` block, where any
  normal exit seals; a nested-record lift test asserts only `shape == (8,)` across three
  dispatches under one fixed seed, so a wrong-column or misaligned-row bug would pass.
- **Count-API edge cases.** The invalid parametrizations for `sample_shape` and the count
  validators omit `np.float64(4.0)`, `jnp.float32(4)`, and out-of-range values such as
  `np.uint64(2**63)`.

## 6. Open design questions

These need a decision rather than a patch.

- **A sanctioned form for temporary provenance-mode changes.** #400 made the mode a per-run
  snapshot on the workflow frame, so a mid-run change is now harmless. There is still no context
  manager for temporary overrides, so every caller mutates the module singleton directly and
  `ProvenanceConfig`'s "Set once at application startup" guidance stays advisory. A
  `with provenance_config.mode_set(...)` would give legitimate temporary changes a defined shape,
  and would be the right place to refuse entry while a run is active.
- **Absolute source paths in provenance diagnostics.**
  `diagnostics["callable_source"]["source_location"]` records an absolute local filesystem path,
  so exported provenance embeds the user's home directory and repository layout. It is documented
  diagnostic content rather than replay authority, so this is a privacy and product call; a
  repo-relative path plus the existing digest would carry the same information.
- **Where Prefect's result store sits relative to the trust boundary.**
  `_assert_transported_frame_consistency` is honestly named — it checks self-consistency, not
  payload authenticity — and the parent still binds returned reports to its own reservation. The
  consequence is that `payload.parent.root_words` is trusted verbatim, so anything able to write
  Prefect's parameter or result store chooses the RNG root for that unit. That is a reasonable
  boundary to accept; it should be written down where the execution contract is described.
